### ISSUE TITLE
Kontrola hodnot v dateissued v číslech a přílohách

### ISSUE THREAD

**Issue by luckajirku** [proarc/proarc-client#399]
Odložím si tu issue s nejasným zadáním zatím k přemýšlení/diskusi. Bylo by fajn, kdyby šlo nějak víc hlídat u periodik datum. Kromě formátu jako takového (je v https://github.com/proarc/proarc-client/issues/378), kdyby PA upozornil na možnou chybu, pokud by číslo (příloha) pod ročníkem neodpovídalo tomu roku, který je v popisu ročníku. Stává se, že v čísle člověk zadá jiný rok (typicky ten současný, nebo na začátku dalšího ročníku dá automaticky rok z ročníku předchozího apod.)... Budou tam i zádrhele jako rozmezí let atd., ale hlídat by se to myslím dalo.

**Comment by ZdenkaSera** [comment]
👍 - Můžeme k tomu dát "návrh na rozšíření"?

**Comment by luckajirku** [comment]
jj, tak to bylo myšleno, aby se s tím počítalo. i proto to sem dávám teď, i když je to takové hodně neurčité zatím...

**Comment by luckajirku** [comment]
Pak je otázka, jestli by to mělo kontrolovat už přímo při ukládání formuláře toho čísla/přílohy (bude fungovat jen když bude objekt ve struktuře, u samostatně založených by neměl proti čemu kontrolovat - ale těch bude minimum). Kde by to dávalo smysl vždycky, to by byla validace na úrovni ročníku, že by to směrem dolů pak zkontrolovalo, jestli roky sedí. Stejně tak u validace při exportu.

**Comment by ZdenkaSera** [comment]
Týká se ručně zakládaných ročníků/čísel/příloh period a e-periodik, tj. ne objektů stahovaných z katalogu

**Comment by luckajirku** [comment]
jako úplný základ by tady měla být kontrola povolených hodnot na té které pozici (v závislosti na použitém formátu -  https://github.com/proarc/proarc-client/issues/378) - př. MM může mít jen hodnotu 01-12 atp (teď můžu zadat klidně třináctý měsíc).

**Comment by luckajirku** [comment]
nebo jestli to půjde napojit na kalendář - a mít u dat obdobné kontroly jako při vytváření čísel periodika přes více...

**Comment by SykoraLukas** [comment]
Při uložení čísla kontrolovat, že rok dateIssue odpovídá roku vydání ročníku.

Kontrola pro NDK Číslo, NDK přílohu čísla, NDK přílohu ročníku.

@kerschfilip ověřit u NDK eDokumentů.

**Comment by SykoraLukas** [comment]
> Týká se ručně zakládaných ročníků/čísel/příloh period a e-periodik, tj. ne objektů stahovaných z katalogu

Toto nejde odlišit. Informaci o tom, že data jsou stažena z katalogu víme jen v době stažení. Poté již ne, takže se bude kontrolovat při každém stažení.

Kontrola tedy bude zda dateIssued odpovídá ročníku, ke kterému je přidán. Vzhledem k tomu, že u určitých okrajových čísel může dojít k tomu, že vyšlo ještě v předchozím nebo následujícím roce, bych udělal toto jako překonatelnou chybu při exportu.


Kontrolovat se kromě toho bude, že je daná hodnota pro den, měsíc a rok v odpovídajícím číselném rozmezí.

**Comment by SykoraLukas** [comment]
Kontrolováno, zda hodnota odpovídá ročníku. Ve validaci se to hlásí jako překonatelná chyba, aby při exportu toto šlo překonat.

**Comment by SykoraLukas** [comment]
Kontroluje se na straně API při uložení.

**Comment by tumovam** [comment]
@luckajirku mohla by ses prosím podívat na toto issue? Děkuji

**Comment by luckajirku** [comment]
při zakládání čísla se jiným rokem, než má ročník, to nedá žádnou hlášku, při spuštění validace na úrovni ročníku taky ne

**Comment by luckajirku** [comment]
ročník 2021, pod kterým jsou čísla z roku 2022: 491c9e33-b633-46a3-9b64-ca0896876412

**Comment by SykoraLukas** [comment]
Vyzkoušeno a nechová se mi to tak, jak popisuješ. Ano při založení se žádná hláška nevyhodí, protože se toto nekontroluje při založení.

Při spuštění ruční validace se ve správě dávek pro každé číslo založí vlastní proces. U každého, co jsi uvedla je podobná hláška:
![image](https://github.com/user-attachments/assets/e1699d33-1f9b-4ab2-b2c3-7164e31c6aef)

Upozornění je to z toho důvodu, že při exportu to je překonatelná chyba. Tato struktura s číslem z jiného roku, než je hodnota v ročníku je akceptovatelná (kvůli číslům, které vycházejí už koncem předchozího nebo ještě začátkem následujícího roku).

I v obrazovce hledat je možnost zobrazit si onu chybovou hlášku:
![image](https://github.com/user-attachments/assets/a31c2622-a87e-46c4-97ec-8342757f5d47)


Prosím ještě jednou o kontrolu.

**Comment by luckajirku** [comment]
popravdě asi nechápu, kde/jak to pouštíš a zobrazíš si ty hlášky. jsem v editaci, spustím v tom ročníku validaci, dostanu tohle:
![validacedateissued](https://github.com/user-attachments/assets/06607adf-18d8-45b4-89fc-12993245ef4d)

**Comment by SykoraLukas** [comment]
Jenže to, co ty pouštíš kontroluje, zda jsou vyplněné všechny pole dle povinnosti M ve formuláři. Kontrola probíhá jen na podřízených objektech.

Já pouštím validaci před exportem zde:
![image](https://github.com/user-attachments/assets/f67f719f-9330-4065-b391-08c52ccc7731)

A zobrazuju si to ve správě procesů:
![image](https://github.com/user-attachments/assets/e2ab370a-a3ec-4025-b103-88b48663ec97)

**Comment by luckajirku** [comment]
aha, ale tak nepokryje celé zadání:-). Jednak jde o kontrolu při kontrolách, a pak další před exportem. viz 
> validace na úrovni ročníku, že by to směrem dolů pak zkontrolovalo, jestli roky sedí. Stejně tak u validace při exportu.

**Comment by SykoraLukas** [comment]
Již se kontroluje i při uložení čísla (za podmínky, že je přiřazeno číslo k ročníku).

Ve validaci před exportem se kontrolovalo dříve.
Ve validaci, kterou si pouštím z obrazovky hledat se kontroloval také dříve.

**Comment by luckajirku** [comment]
Jenomže teď když je při zakládání ta chyba jako nepřekonatelná, tak nezaložím vůbec ta "hraniční" čísla toho ročníku - vydaná třeba v lednu následujícího nebo prosinci předchozího roku... 
Potřebujeme, aby to bylo překonatelné, a hlavně aby ta validace fungovala v editaci, tam probíhají primárně kontroly a hned se tam ty chyby i opravují (v obrazovce hledat to je fajn, ale hlavně to potřebujeme v editaci).

**Comment by SykoraLukas** [comment]
Zakládání je opravené.

O přidání funkce na spouštění validace "před exportem" prosím @albertoh. Ale výsledek té funkce stejně je, že se to zapisuje do procesů.

**Comment by tumovam** [comment]
@SykoraLukas nevím, zda něco nedělám špatně, ale když mám ročník a pod ním dvě čísla, kdy jedno vyšlo ve stejném roce jako ročník a druhé číslo vyšlo v následujícím roce, tak export neproběhne. Zdá se mi tedy, že nejde o překonatelnout chybu. Ve Správě procesů upozornění: _"Objekt nemá validní dateIssued vůči svému nadřazenému objektu"_ . Můžeš se prosím podívat, zda se za rok otevření toho issue něco nezměnilo? Předem děkuji

**Comment by SykoraLukas** [comment]
Na klientovi je přidána funkce pod tři tečky v `Editoru vazeb`.

![Image](https://github.com/user-attachments/assets/78d1a394-f3b4-471a-b7fe-89f6822b1de5)

**Comment by SykoraLukas** [comment]
K testování.

**Comment by tumovam** [comment]
Podařilo se mi vyexportovat (Kramerius export) dvě čísla s eČlánky - jedno 01.2025, druhé 06.2026, které jsou pod ročníkem 2024. Za nás tedy v pořádku 🙂

**Comment by tumovam** [comment]
NDK Přílohu periodika nelze vyexportovat (test pouze Kramerius export), pokud vyšla v jiném roce, než ročník. @luckajirku napadá Tě prosím případ, kdy příloha vyšla v jiný rok, než ročník? Díky

**Comment by tumovam** [comment]
@SykoraLukas ano, já bych byla pro překonatelnou chybu jako u NDK Čísla. Díky

**Comment by SykoraLukas** [comment]
Na jakém objektu to prosím zkoušíš? Ať to testujeme na stejném objektu.

**Comment by tumovam** [comment]
Jo, byly to přílohy 8ff634cd-06c4-4ca2-9299-58b9adf57607 (jiný rok vyd.) a fc533e31-2e76-449f-9fc7-46a32b28374e (tam je stejný rok vyd.).

**Comment by SykoraLukas** [comment]
Ano, ale tam to nepadá kvůli roku vydání, ale kvůli chybě v souvislosti s https://github.com/proarc/proarc/issues/1756.

**Comment by SykoraLukas** [comment]
Je to překonatelná chyba - proto na začátku je "Upozornění", nepřekonatelná chyba začíná na "Chyba"

**Comment by tumovam** [comment]
Překonatelná chyba (upozornění) když je dateIssued jiné vůči svému nadřazenému objektu. Zkoušela jsem např. i 13. měsíc nebo 40. den, příp. různě jiné nepodporované hodnoty a nic mi neprošlo. Myslím, že kontrola hodnot v číslech a přílohách už je v pořádku. Ale prosím o vyjádření a otestování i další zainteresované. Děkuji

**Comment by luckajirku** [comment]
za nás OK, odmítá 13. měsíc, 32. den, nadbytečné mezery, dvě tečky místo jedné atd.

**Comment by katerinasirova** [comment]
Opravdu se mi podařilo uložit do dateIssued datum 30.2.1991 a ani jedna validace se nad tím nepozastavila - resp. nevyhodnotila jako chybu. uuid čísla: 9bee4cbd-428a-4da8-a0e6-ff59141d4174.

![Image](https://github.com/user-attachments/assets/1261b867-796b-4f44-a114-44da1de8d574)

**Comment by SykoraLukas** [comment]
Opraveno, k testování.

**Comment by tumovam** [comment]
@luckajirku @katerinasirova prosím o přetestování. Předem děkuji

**Comment by SykoraLukas** [comment]
Opraveno.
